mithril-stm/src/multi_sig.rs

```diff
@@ -589,6 +535,95 @@ impl_serde!(VerificationKey, VerificationKeyVisitor, 96);
 impl_serde!(ProofOfPossession, ProofOfPossessionVisitor, 96);
 impl_serde!(Signature, SignatureVisitor, 48);
 
+// ---------------------------------------------------------------------
+// Unsafe helpers
+// ---------------------------------------------------------------------
+mod unsafe_helpers {
+    use super::*;
+    use blst::{
+        blst_fp12, blst_fp12_finalverify, blst_p1_affine_generator, blst_p2_affine_generator,
+        blst_sk_to_pk_in_g1,
+    };
+
+    pub(crate) fn verify_miller_loop(vk: &BlstVk, k2: &blst_p1) -> bool {
+        unsafe {
+            let g1_p = *blst_p1_affine_generator();
+            let mvk_p = std::mem::transmute::<&BlstVk, &blst_p2_affine>(vk);
+            let ml_lhs = blst_fp12::miller_loop(mvk_p, &g1_p);
+
+            let mut k2_p = blst_p1_affine::default();
+            blst_p1_to_affine(&mut k2_p, k2);
+            let g2_p = *blst_p2_affine_generator();
+            let ml_rhs = blst_fp12::miller_loop(&g2_p, &k2_p);
+
+            blst_fp12_finalverify(&ml_lhs, &ml_rhs)
+        }
+    }
+
+    pub(crate) fn compress_p1(k2: &blst_p1) -> [u8; 48] {
+        let mut bytes = [0u8; 48];
+        unsafe { blst_p1_compress(bytes.as_mut_ptr(), k2) }
+        bytes
+    }
+
+    pub(crate) fn uncompress_p1(bytes: &[u8]) -> blst_p1 {
+        unsafe {
+            let mut point = blst_p1_affine::default();
+            let mut out = blst_p1::default();
+            blst_p1_uncompress(&mut point, bytes.as_ptr());
+            blst_p1_from_affine(&mut out, &point);
+            out
+        }
+    }
+
+    pub(crate) fn scalar_to_pk_in_g1(sk: &SigningKey) -> blst_p1 {
+        unsafe {
+            let sk_scalar = std::mem::transmute::<&BlstSk, &blst_scalar>(&sk.0);
+            let mut out = blst_p1::default();
+            blst_sk_to_pk_in_g1(&mut out, sk_scalar);
+            out
+        }
+    }
+
+    pub(crate) fn vk_from_p2_affine(vk: &VerificationKey) -> blst_p2 {
+        unsafe {
+            let mut projective_p2 = blst_p2::default();
+            blst_p2_from_affine(
+                &mut projective_p2,
+                &std::mem::transmute::<BlstVk, blst_p2_affine>(vk.0),
+            );
+            projective_p2
+        }
+    }
+
+    pub(crate) fn sig_to_p1(sig: &BlstSig) -> blst_p1 {
+        unsafe {
+            let mut projective_p1 = blst_p1::default();
+            blst_p1_from_affine(
+                &mut projective_p1,
+                &std::mem::transmute::<BlstSig, blst_p1_affine>(*sig),
+            );
+            projective_p1
+        }
+    }
+
+    pub(crate) fn p2_affine_to_vk(grouped_vks: &p2_affines, scalars: &[u8]) -> BlstVk {
```

Comment:
This function signature is very confusing. It's name `p2_affine_to_vk` but you give as input `&p2_affines` and `&[u8]`, and you output a `BlstVk`. This function should already take as input the result of `&grouped_vks.mult(scalars, 128)`, or have a different name (if possible, I'd suggest the former).